src/main/java/executor/service/service/impl/listener/ScenarioSourceListenerImpl.java

```diff
@@ -0,0 +1,83 @@
+import executor.service.service.ScenarioSourceListener;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import reactor.core.publisher.Flux;
+
+import java.time.Duration;
+import java.util.List;
+import java.util.Properties;
+
+import static executor.service.config.properties.PropertiesConstants.*;
+
+/**
+ * The {@code ScenarioSourceListenerImpl} class implements the {@link ScenarioSourceListener} interface
+ * that reads scenarios from a {@link ScenarioProvider}
+ * and emits them as a {@link Flux} stream with a specified delay.
+ * <p>
+ *
+ * @author Yurii Kotsiuba, Oleksandr Tuleninov, Oleksii Bondarenko, Kostia Hromovii
+ * @version 01
+ * @see executor.service.service.ScenarioProvider
+ * @see executor.service.config.properties.PropertiesConfig
+ */
+public class ScenarioSourceListenerImpl implements ScenarioSourceListener {
+
+    private static final Logger log = LoggerFactory.getLogger(ScenarioSourceListener.class);
+
+    private final ScenarioProvider provider;
+    private final PropertiesConfig propertiesConfig;
+
+    public ScenarioSourceListenerImpl(ScenarioProvider provider,
+                                      PropertiesConfig propertiesConfig) {
+        this.provider = provider;
+        this.propertiesConfig = propertiesConfig;
+    }
+
+    /**
+     * Executes the scenario source listener by reading scenarios from the {@link ScenarioProvider},
+     * validating them, and emitting them as a continuous stream using a {@link Flux}.
+     *
+     * @param handler The {@link ItemHandler} to handle received scenarios.
+     */
+    @Override
+    public void execute(ItemHandler handler) {
+        List<Scenario> scenarios = provider.readScenarios(FILE_NAME_SCENARIOS);
+        validateScenarios(scenarios);
+        Flux<Scenario> scenariosFlux = getScenarioFlux(scenarios);
+        scenariosFlux.subscribe(handler::onItemReceived);
+    }
+
+    /**
+     * Check the list of ProxyConfigHolder entities.
+     *
+     * @param scenarios list of Scenario entities
+     */
+    private void validateScenarios(List<Scenario> scenarios) {
+        if (scenarios == null || scenarios.isEmpty()) {
+            log.error("The scenarios list is bad.");
+        }
+    }
+
+    /**
+     * Retrieves a {@link Flux} stream with {@link Scenario} entities continuously, applying a delay between emissions.
+     *
+     * @param scenarios The list of {@link Scenario} entities to emit.
+     * @return A {@link Flux} stream emitting scenarios with the specified delay.
+     */
+    private Flux<Scenario> getScenarioFlux(List<Scenario> scenarios) {
+        return Flux.fromIterable(scenarios)
+                .delayElements(Duration.ofSeconds(getDelay()))
+                .repeat();
+    }
+
+    private Long getDelay() {
+        Properties properties = propertiesConfig.getProperties(SOURCES_PROPERTIES);
+        return Long.parseLong(properties.getProperty(DELAY_SCENARIO_SECONDS));
+    }
```

Comment:
This is fantastic, every call your decide to read the same value from properties?  It's too hard set value on create and then use it common?

Reply:
The delayElements(Duration delay) method uses the specified Duration to delay each signal (Subscriber#onNext) in the stream. The delay occurs on a parallel plan (parallel) using the scheduler (Scheduler), which by default is used for parallel operations (Schedulers.parallel()).

This method allows you to create a new Flux in which each signal will be delayed for a specified amount of time, but this will not apply to empty sequences or spurious signals that occur immediately.
![image](https://github.com/GFLCourses5/worker/assets/140727960/0b128185-7e38-4159-a52b-830e247bf481)
You can check it in **package reactor.core.publisher;** - **public abstract class Flux<T>**.

Reply:
> e delayElements(Duration delay) method uses the specified Duration to delay each signal (Subscriber#onNext) in the stream. The delay occurs on a parallel plan (parallel) using the scheduler (Scheduler), which by default is used for parallel operations (Schedulers.parallel()).

How *delayElements(Duration delay)* is related to this question?

Reply:
Anton suggested that with each call the same value is read from the properties. And this is not true. Flux is created and initialized once.
![image](https://github.com/GFLCourses5/worker/assets/140727960/4a987910-380e-442b-8970-c92b0d8c8fd0)

Reply:
> Anton suggested that with each call the same value is read from the properties. And this is not true. Flux is created and initialized once. ![image](https://user-images.githubusercontent.com/140727960/274180087-4a987910-380e-442b-8970-c92b0d8c8fd0.png)

Question was about *getDelay()* method, not the flux. If for some reason you need to call it somewhere else within your class, you will be forced to read the same value from properties file again. Thats fine that now you have only 1 call and thats ok, but potentially this approach can bring troubles.